### DISCUSSION TITLE
Thumbnails and generative AI

### DISCUSSION THREAD

**Discussion by kdill00** [blakeblackshear/frigate#14068] [Ask A Question] [answered]
Hey guys. Ive been playing a bit with the semantic and generative AI and im curious if it is possible to use snapshots for generating texts?  The quality of thumbnails seems to cause wrong outputs in some cases, especially objects in a higher motion. With higher resolution and quality snapshots, it can accurately lift plate numbers from my front ptz camera. The thumbnails seem to blur it too much.

**Comment by NickM-27** [comment]
CPU usage can already get quite high when running embeddings on objects like cars in even medium traffic areas, this greatly increases that usage even further.

**Reply by kdill00** [reply]
Could it only be performed at the end, on potentially the best snapshot?  Or some kind of better targeting of what images are submitted?  Could the generative one only use snapshots?  The cpu based ones i understand keeping it to thumbnails.  Clip and those work surprisingly well for the quality of thumbnails. Im more asking about the generative. I have an 80gb nvidia a100 and it is working through all the requests with room to spare.  Maybe a scheduling type deal where you could have it return to process at certain times automatically when resources are more available?  I see great potential with these features. Just picking your brain with where you see it headed.

**Reply by NickM-27** [reply]
It still adds complication given that different models have different max dimensions for the image that they can run on, and that will add complication as well as a lot more error handling being required. I did not contribute this feature nor am I really working on it personally so I will defer to @hawkeye217 and @hunterjm if they have specific suggestions

**Reply by hawkeye217** [reply, marked as answer]
I've implemented using the snapshot for genai here: https://github.com/blakeblackshear/frigate/pull/14077

Please test it and let me know if it's working well for you.

**Reply by kdill00** [reply]
Thanks guys. Ill give it a go and let you know.   Appreciate the rapid turn around

**Reply by kdill00** [reply]
In the few days testing it, I think I can see benefit to both. With snapshots, the model stops complaining about blurriness as often but it does say sometimes it lacks enough information from a single image to make an accurate call. So when its feeding thumbnails, the model is generating a response on multiple images and with snapshots when the event end is fired, the best image is submitted?  Also it may be better if possible to make it opt in instead of out. I had to define the objects on each camera as none to prevent submission on all 40 cameras at one location. I dont know if that possible. Just a thought, always looking to slim down a thousand line config. Will keep playing with it.

**Reply by kdill00** [reply]
If the number and frequency of images could be tuned in the config that are sent per generative event, that would be nice.  I am only employing it on cameras covering our entrances and exits, 2 of them. As our business caters to younger kids(amusement park), i always want to be able provide everything I can in instances we are called upon.  I would max out any kind of settings that increase accuracy no matter the cost to resources. Im thinking of running lighter models but clip and the other embeddings actually do a decent job without the generative ai on cameras that dont need as intense watching.  Cheers guys. This beta has really pushed it into a new realm much like when blake made the changes to constantly track detections.

**Reply by kdill00** [reply]
I am an extreme use case and I do understand this. So do with this information what you will!

**Reply by hawkeye217** [reply]
Glad it's working well for you!

To be clear, the snapshot submission is opt-in (see the dev docs [here](https://deploy-preview-13787--frigate-docs.netlify.app/configuration/genai#custom-prompts)). By default, the smaller thumbnails over the life of the tracked object are sent to the model in a single request.
